hazardous/tests/test_ipcw.py

```diff
@@ -0,0 +1,52 @@
+import numpy as np
+import pytest
+from lifelines.datasets import load_regression_dataset
+from numpy.testing import assert_array_almost_equal
+
+from .._ipcw import IpcwEstimator
+
+
+@pytest.mark.parametrize("competitive_risk", [True, False])
+def test_ipcw(competitive_risk):
+    X = load_regression_dataset()
+    y = dict(
+        event=X["E"],
+        duration=X["T"],
+    )
+    if competitive_risk:
+        coef = np.random.choice([1, 2], size=y["event"].shape[0])
+        y["event"] *= coef
+
+    times = np.arange(
+        y["duration"].min(),
+        y["duration"].max() - 1,
+    )
+
+    est = IpcwEstimator().fit(y)
+    ipcw_probs = est.predict(times)
+
+    expected_ipcw_probs = np.array(
+        [
+            1.0,
+            1.0,
+            1.0,
+            1.01639935,
+            1.02308619,
+            1.03024064,
+            1.03816557,
+            1.03816557,
+            1.06416786,
+            1.06416786,
+            1.08835349,
+            1.08835349,
+            1.08835349,
+            1.19718884,
+            1.33020982,
+            1.33020982,
+            1.33020982,
+            1.33020982,
+            1.33020982,
+        ]
+    )
+
+    assert_array_almost_equal(ipcw_probs, expected_ipcw_probs)
```

Comment:
Instead of checking specific values that come from nowhere (but our current implementation), let's instead check expected mathematical properties:

```suggestion
    # Probabilities lie in [0.0, 1.0] hence the range of their inverse
    # is [1.0, +inf).
    assert (ipcw_probs >= 1.0).all(), ipcw_probs
    
    # Expect monotonically increasing censoring probabilities over time,
    # therefore monotonically increasing IPCW values.
    assert (ipcw_probs.diff() >= 0).all(), ipcw_probs
```

Also, nowadays, `assert_allclose` is considered a better replacement to `assert_array_almost_equal` (because the way tolerance is defined makes more sense usually).

Reply:
Actually I am confused now. Are IPCW "inverse probabilities of censoring" or "inverse probabilities of absence of censoring"?

Reply:
This values don't come from nowhere, they are the reference from scikit-survival and are far more reliable than some macro statistical checks.

Reply:
IPCW = Inverse Probability of Censoring Weight (your first guess), we compute these with Kaplan Meier by considering censoring as the event of interest.

Reply:
> This values don't come from nowhere, they are the reference from scikit-survival and are far more reliable than some macro statistical checks.

Alright, then let's make this explicit in the code that makes it possible to reproduce those values with scikit-survival by adding the code snippet in commented out lines.

Reply:
> IPCW = Inverse Probability of Censoring Weight (your first guess), we compute these with Kaplan Meier by considering censoring as the event of interest.

Now I realize that this name (IPCW) is a bit misleading: 

KM on `(1 - any_event_indicator)` indeed estimates the **survival to censoring**. Hence, it estimates the cumulative probability of "not-censoring":

At t=0, the probability of censoring is always 0, and IPCW is typically at its lowest value, close to 1.0. When time progresses, the cumulative probability of censoring can only increases. The probability of its complement (probability of not censoring) monotically decreases and IPCW therefore monotonically increases (and always stays above 1.0 by construction).

On some datasets, the censoring probability reach 1.0 after t_max (all datapoints recorded with t > t_max are censored). In this case, the probability of not being censored is getting very close to 0 and IPCW can go to +inf.

I will try to make a suggestion to add a comment to make that explicit in the code.

Reply:
I updated the code, docstring and tests to take those remarks into account.